File: localsearch/MCMF.cpp
#include <math.h>
#include <vector>
#include <queue>

using namespace std;

#define ZERO 1e-8

class EDGE {
	public:
		int v, next;   //----------------------cap,cost: int?
		double cap,cost;
		int original;
		EDGE()
		:v(0),
		 next(0),
		 cap(0.0),
		 cost(0.0),
		 original(0.0) {}
	
		EDGE(int a, double b, double c,int d,int e)
		:v(a),
		 cap(b),
		 cost(c),
		 next(d),
		 original(e) {}
};

class MCMF {
	public:
		constexpr static double inf = 2000000.0;
	
		vector<int> from;
		vector<int> to;
		vector<double> flow;

		void init(int _n){
			n = _n;
			size = 0;
			vis.reserve(n+1);
			net.reserve(n+1);
			for(int i=0;i<=n;i++)
				net[i] = -1;
			pre.reserve(n+1);
			cur.reserve(n+1);
			dis.reserve(n+1);
		}
	
		/**
		 * u & v are source vertex and destination vertex
		 * cap are the edge capacity
		 * cost are the edge cost of a unit flow
		 */
		void AddEdge(int u, int v, double cap, double cost){
			E.emplace_back(v, cap, cost, net[u], 1);
			net[u] = size++;
			E.emplace_back(u, 0, -cost, net[v], 0);
			net[v] = size++;	
		}

		double calMCMF(int s, int t){
			st = s;
			en = t;
			spfa();
			mincost = maxflow = 0;
			for(int i = 0 ; i <= n; i++){
				vis[i] = false;
				cur[i] = net[i];
			}
			do{
				while(fabs(augment(st,inf)) > ZERO){
					for(int i = 0 ; i <vis.size(); i++){
						vis[i] = false;	
					}
				}		
			}while(adjust());
			transOutput();
			return mincost;
		}

		double MincostMaxflow(int s,int t){
			return calMCMF(s,t);
		}

	private:
		int size, n;
		int st, en;
		double maxflow, mincost;
		vector<bool> vis;
		vector<int> net, pre, cur;
		vector<double> dis;
		vector<EDGE> E;
		bool flag;
	
		queue<int> Q;
	
		bool adjust(){
			int v;
			double min = inf;
			for(int i = 0; i <= n; i++){
				if(!vis[i])
					continue;
				for(int j = net[i]; j != -1 ; j = E[j].next){
					v = E[j].v;
					if(fabs(E[j].cap) > ZERO)
						flag = true;
					else
						flag = false;
					if(flag)
						if(!vis[v] && dis[v]-dis[i]+E[j].cost < min){
							min = dis[v] - dis[i] + E[j].cost;
						}
				}
			}
			if(fabs(min - inf) < ZERO)
				return false;
			for(int i = 0 ; i <= n; i++)
				if(vis[i]){
					cur[i] = net[i];
					vis[i] = false;
					dis[i] = min+dis[i];
				}
			return true;
		}
	
		double augment(int i, double flow){
			if(i == en){
				mincost += dis[st] * flow;
				maxflow += flow;
				return flow;
			}
			vis[i] = true;
			for(int j = cur[i],v ; j != -1 ;j = E[j].next){
				v = E[j].v;
				if(fabs(E[j].cap) < ZERO)
					continue;
				if(vis[v] || fabs(dis[v]+E[j].cost - dis[i]) > ZERO)
					continue;
				double delta = augment(v, flow < E[j].cap ? flow : E[j].cap);
				if(fabs(delta) > ZERO){
					E[j].cap -= delta;
					E[j^1].cap += delta;
					cur[i] = j;
					return delta;
				}
			}
			return 0;
		}
	
		void spfa(){
			int u,v;
			for(int i = 0; i <=n ; i++){
				vis[i] = false;
				dis[i] = inf;
			}
			dis[st] = 0;
			Q.emplace(st);
			vis[st] = true;
			while(!Q.empty()){
				u = Q.front();
				Q.pop();
				vis[u] = false;
				for(int i = net[u]; i != -1; i = E[i].next){
					v = E[i].v;
					if(fabs(E[i].cap) < ZERO||dis[v] <= dis[u]+E[i].cost)
						continue;
						
					dis[v] = dis[u]+E[i].cost;
					if(!vis[v]){
						vis[v] = true;
						Q.emplace(v);
					}
				}
			}
			for(int i = 0 ; i <= n ;i++)
				dis[i] = dis[en]-dis[i];
		}
	
		void transOutput(){
	    for(int i = 0 ; i <= n ; i++){
	        int nodeId = i;
	        int edgeId = net[nodeId];
	        while(edgeId!=-1){
	            int rightId = E[edgeId].v;
	            double edgeflow = inf - E[edgeId].cap;
	            if(E[edgeId].original == 1 && fabs(edgeflow)>ZERO){
	                from.emplace_back(nodeId);
	                to.emplace_back(rightId);
	                flow.emplace_back(edgeflow);
	            }
	            edgeId = E[edgeId].next;
	        }
	    }
		}
};
/*
int main() {
		MCMF augraph;     
		augraph.init(5); 
		
		augraph.AddEdge(1 ,2 ,10, 4);
		augraph.AddEdge(1 ,3 ,MCMF::inf, 1);
		augraph.AddEdge(3 ,2 ,5, 2);
		augraph.AddEdge(3 ,4 ,10,3);
		augraph.AddEdge(2 ,4 ,6, 2);
		augraph.AddEdge(2 ,5 ,7, 1);
		augraph.AddEdge(4 ,5 ,4, 2);
		
		augraph.MincostMaxflow(1, 5);
		return 0;
}
*/

File: appro/MCMF.java
package appro;

import java.util.*;
import java.io.*;
import java.lang.*;

public class MCMF {
	public static final double inf = 2000000;
	
	public List<Integer> from;
	public List<Integer> to;
	public List<Double> flow;
	
	int size, n;
	int st, en;
	double maxflow, mincost;
	boolean[] vis;
	int[] net, pre, cur;
	double[] dis;
	ArrayList<EDGE> E;
	boolean flag;
	
	Queue<Integer> Q;
	
	public void init(int _n){
		n = _n;
		size = 0;
		vis = new boolean[n+1];
		net = new int[n+1];
		for(int i=0;i<=n;i++)
			net[i] = -1;
		pre = new int[n+1];
		cur = new int[n+1];
		dis = new double[n+1];
		E = new ArrayList<EDGE>();
		Q = new ArrayDeque<Integer>();
        from = new ArrayList<Integer>();
        to = new ArrayList<Integer>();
        flow = new ArrayList<Double>();
	}
	
	public void AddEdge(int u, int v, double cap, double cost){
		E.add(new EDGE(v, cap, cost, net[u], 1));
		net[u] = size++;
		E.add(new EDGE(u, 0, -cost, net[v], 0));
		net[v] = size++;	
	}
	
	private boolean adjust(){
		//System.exit(0);
		int v;
		double min = inf;
		for(int i = 0; i <= n; i++){
			if(!vis[i])
				continue;
			for(int j = net[i]; j != -1 ; j = E.get(j).next){
				v = E.get(j).v;
				if(Math.abs(E.get(j).cap) > 1e-8)
					flag = true;
				else
					flag = false;
				if(flag)
					if(!vis[v] && dis[v]-dis[i]+E.get(j).cost < min){
						min = dis[v] - dis[i] + E.get(j).cost;
					}
			}
		}
		if(Math.abs(min - inf) < 1e-8)
			return false;
		for(int i = 0 ; i <= n; i++)
			if(vis[i]){
				cur[i] = net[i];
				vis[i] = false;
				dis[i] = min+dis[i];
			}
		//System.out.println(min);
		return true;
	}
	
	private double augment(int i, double flow){
		if(i == en){
			mincost += dis[st] * flow;
			maxflow += flow;
			return flow;
		}
		vis[i] = true;
		for(int j = cur[i],v ; j != -1 ;j = E.get(j).next){
			v = E.get(j).v;
			if(Math.abs(E.get(j).cap) < 1e-8)
				continue;
			if(vis[v] || Math.abs(dis[v]+E.get(j).cost - dis[i]) > 1e-8)
				continue;
			double delta = augment(v, Math.min(flow, E.get(j).cap));
			if(Math.abs(delta) > 1e-8){
				E.get(j).cap -= delta;
				E.get(j^1).cap += delta;
				cur[i] = j;
				return delta;
			}
		}
		return 0;
	}
	
	private void spfa(){
		//System.out.println("spfa start");
		int u,v;
		for(int i = 0; i <=n ; i++){
			vis[i] = false;
			dis[i] = inf;
		}
		dis[st] = 0;
		Q.add(st);
		vis[st] = true;
		while(!Q.isEmpty()){
			u = Q.poll();
			vis[u] = false;
			for(int i = net[u]; i != -1; i = E.get(i).next){
				v = E.get(i).v;
				if(Math.abs(E.get(i).cap) < 1e-8||dis[v] <= dis[u]+E.get(i).cost)
					continue;
					
				dis[v] = dis[u]+E.get(i).cost;
				if(!vis[v]){
					vis[v] = true;
					Q.add(v);
				}
			}
		}
		for(int i = 0 ; i <= n ;i++)
			dis[i] = dis[en]-dis[i];
		//System.out.println("spfa done");
	}
	
	private void transOutput(){
	    for(int i = 0 ; i <= n ; i++){
	        int nodeId = i;
	        int edgeId = net[nodeId];
	        while(edgeId!=-1){
	            int rightId = E.get(edgeId).v;
	            double edgeflow = inf - E.get(edgeId).cap;
	            if(E.get(edgeId).original == 1 && Math.abs(edgeflow)>1e-8){
	                from.add(nodeId);
	                to.add(rightId);
	                flow.add(edgeflow);
	                //System.out.println(nodeId + ":" + rightId + ":" + edgeflow);
	            }
	            edgeId = E.get(edgeId).next;
	        }
	    }
	}
	
	public double calMCMF(int s, int t){
		st = s;
		en = t;
		spfa();
		mincost = maxflow = 0;
		//System.out.println("calMCMF start");
		for(int i = 0 ; i <= n; i++){
			vis[i] = false;
			cur[i] = net[i];
		}
		do{
			while(Math.abs(augment(st,inf)) > 1e-8){
				for(int i = 0 ; i <vis.length; i++){
					vis[i] = false;	
				}
			}		
		}while(adjust());
		//System.out.println("calMCMF done!");
		transOutput();
		return mincost;
	}
	
	public double MincostMaxflow(int s,int t){
		return calMCMF(s,t);
	}
	
	public static void main(String arg[]){
		MCMF augraph = new MCMF();      
		augraph.init(5); 
		
		augraph.AddEdge(1 ,2 ,10, 4);
		augraph.AddEdge(1 ,3 ,8, 1);
		augraph.AddEdge(3 ,2 ,5, 2);
		augraph.AddEdge(3 ,4 ,10,3);
		augraph.AddEdge(2 ,4 ,6, 2);
		augraph.AddEdge(2 ,5 ,7, 1);
		augraph.AddEdge(4 ,5 ,4, 2);
		
		System.out.println(augraph.MincostMaxflow(1, 5));
		
	}

}